processor/src/fast/mod.rs

```diff
@@ -0,0 +1,324 @@
+            Operation::End => todo!(),
+            Operation::Repeat => todo!(),
+            Operation::Respan => todo!(),
+            Operation::Halt => todo!(),
+            Operation::Add => match self.stack.len() {
+                0 | 1 => (),
+                _ => {
+                    let top = self.stack.pop().unwrap_or(ZERO);
+                    *self.stack.last_mut().unwrap() += top;
+                },
+            },
+            Operation::Neg => todo!(),
+            Operation::Mul => todo!(),
+            Operation::Inv => todo!(),
+            Operation::Incr => todo!(),
+            Operation::And => todo!(),
+            Operation::Or => todo!(),
+            Operation::Not => todo!(),
+            Operation::Eq => todo!(),
+            Operation::Eqz => todo!(),
+            Operation::Expacc => todo!(),
+            Operation::Ext2Mul => todo!(),
+            Operation::U32split => todo!(),
+            Operation::U32add => todo!(),
+            Operation::U32assert2(_) => todo!(),
+            Operation::U32add3 => todo!(),
+            Operation::U32sub => todo!(),
+            Operation::U32mul => todo!(),
+            Operation::U32madd => todo!(),
+            Operation::U32div => todo!(),
+            Operation::U32and => todo!(),
+            Operation::U32xor => todo!(),
+            Operation::Pad => todo!(),
+            Operation::Drop => todo!(),
+            Operation::Dup0 => match self.stack.last() {
+                Some(&last_element) => self.stack.push(last_element),
+                // stack is empty, so we're duplicating ZERO
+                None => (),
+            },
+            Operation::Dup1 => match self.stack.len() {
+                0 => (),
+                1 => self.stack.push(ZERO),
+                _ => {
+                    let to_dup_index = self.stack.len() - 2;
+                    self.stack.push(self.stack[to_dup_index]);
+                },
+            },
+            Operation::Dup2 => todo!(),
+            Operation::Dup3 => todo!(),
+            Operation::Dup4 => todo!(),
+            Operation::Dup5 => todo!(),
+            Operation::Dup6 => todo!(),
+            Operation::Dup7 => todo!(),
+            Operation::Dup9 => todo!(),
+            Operation::Dup11 => todo!(),
+            Operation::Dup13 => todo!(),
+            Operation::Dup15 => todo!(),
+            Operation::Swap => {
+                match self.stack.len() {
+                    // We're swapping ZERO with ZERO, which is a no-op
+                    0 => (),
+                    // the second element on the stack is implicitly ZERO, so swapping puts a
+                    // ZERO on top
+                    1 => self.stack.push(ZERO),
+                    _ => {
+                        let last = self.stack.len() - 1;
+                        // TODO(plafer): try swap_unchecked
+                        self.stack.swap(last, last - 1);
+                    },
+                }
```

Comment:
I wonder if there is a way to avoid handling different cases for the current stack length. Maybe we pre-allocate a vector with a bunch of zeros and would always have more than 16 values on the stack. The, we could reduce this to something like:
```rust
let last = self.stack.len() - 1;
let ptr = self.stack.as_mut_ptr();
unsafe {
    std::ptr::swap(ptr.add(last), ptr.add(last - 1));
}
```
Which should have something like 10x smaller cycle count than the current code.